backend: make crawlconfigs mutable! (#656)

ikreymer opened the pull request:
- crawlconfig PATCH /{id} can now receive a new JSON config to replace the old one (in addition to scale, schedule, tags)
- exclusions: add / remove APIs mutate the current crawlconfig, do not result in a new crawlconfig created
- k8s: crawlconfig json is updated along with scale
- k8s: stateful set is restarted by updating annotation, instead of changing template
- crawl object: now has 'config', as well as 'profileid', 'schedule', and 'crawlTimeout' properties to ensure anything that is changeable is stored on the crawl
- crawlconfig object: remove 'oldId', 'newId', disallow deactivating/deleting while crawl is running

Still do:
This change will require a DB migration, to remove crawlconfig props, and populate new props on the crawl

ikreymer committed:
backend: make crawlconfigs mutable! (#656)
- crawlconfig PATCH /{id} can now receive a new JSON config to replace the old one (in addition to scale, schedule, tags)
- exclusions: add / remove APIs mutate the current crawlconfig, do not result in a new crawlconfig created
- k8s: crawlconfig json is updated along with scale
- k8s: stateful set is restarted by updating annotation, instead of changing template
- crawl object: now has 'config', as well as 'profileid', 'schedule', and 'crawlTimeout' properties to ensure anything that is changeable
is stored on the crawl
- crawlconfig object: remove 'oldId', 'newId', disallow deactivating/deleting while crawl is running

ikreymer committed:
reformat

ikreymer committed:
fix setting scale, format

ikreymer committed:
add revision tracking to crawlconfig:
- store each revision separate mongo db collection
- revisions accessible via /crawlconfigs/{cid}/revs
- update 'userid' on each crawlconfig modification, use 'modified' for modification date, 'useridCreated' for initial creation user
- store 'rev' int in crawlconfig and in crawljob
- backwards compatibility: add_crawl_config calls update_crawl_config and returns same id if oldId provided

ikreymer committed:
lint fix

ikreymer committed:
crawlconfigs: only add revision history if crawl data changed, not metadata (tags, name)
tests: add tests for updating config, revision history api

tw4l commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -552,7 +601,7 @@ async def do_make_inactive(self, crawlconfig: CrawlConfig):
 
-    async def copy_add_remove_exclusion(self, regex, cid, org, user, add=True):
+    async def add_remove_exclusion(self, regex, cid, org, user, add=True):
         """create a copy of existing crawl config, with added exclusion regex"""
```
Might be good to update this docstring

tw4l reviewed (commented):
Haven't had a chance to test yet, but the code looks good! Left one small comment about a docstring that could be updated

ikreymer committed:
backend: ensure crawl job 'config' is updated when exclusions are added/removed, unify add/remove exclusions on crawl
remove unused 'completions' field

tw4l committed:
Add migration for mutable crawl configs

tw4l commented:
I tested the migration manually by creating some crawls and crawl configs on version `0002` of the database, then upgrading via helm and testing that all looked good after the migration.

tw4l commented:
@ikreymer I've now tested migrations from an empty slate and from existing 0002 data and all is looking good in terms of being able to re-run configs after migration and create new configs/crawls. API returns what I'd expect as well.

tw4l committed:
Add crawlTimeout to UpdateCrawlConfig and add test

SuaYoo committed:
Use new crawl config update API (#672)

tw4l committed:
WIP: Rename userid fields to createdBy and modifiedBy

ikreymer commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -238,17 +275,29 @@ async def add_crawl_config(
         """Add new crawl config"""
+
+        # for now, to support frontend update logic
+        if config.oldId:
```
I guess we can just remove this conditional now?

tw4l replied:
removed

ikreymer commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -108,9 +108,31 @@ class CrawlConfigIn(BaseModel):
     scale: Optional[conint(ge=1, le=MAX_CRAWL_SCALE)] = 1
 
+    # for now, until frontend is changed
     oldId: Optional[UUID4]
```
I guess can remove this now with frontend changes

tw4l replied:
removed

ikreymer commented on backend/btrixcloud/crawls.py:
```diff
@@ -602,49 +613,41 @@ async def get_redis(self, crawl_id):
 
-        raw = await self.get_crawl_raw(crawl_id, org)
+        # pylint: disable=fixme
+        # todo: remove new_cid once frontend is updated
```
Can remove now

tw4l committed:
Linting fixups

ikreymer committed:
migration: skip invalid crawls (missing config), make createdBy optional (just in case)

ikreymer committed:
lint fix

ikreymer committed:
more userid -> modifiedBy renamings

SuaYoo committed:
Update crawl config keys with new API (#681)

tw4l committed:
Remove oldId

tw4l committed:
Remove new_id from response

tw4l committed:
Add missing return to fix /run response

SuaYoo committed:
update config object

SuaYoo committed:
update crawlTimeout type

SuaYoo committed:
update crawlTimeout in initial crawl config

ikreymer committed:
backend:
- ensure 'jobType' also copied to crawl
- ensure 'profileName' is resolved on CrawlOut from profileid
- add shared CrawlConfigCore which shared properties set on both crawlconfig (soon: workflow) and crawl
- migration: update migration to also copy jobType, force run

ikreymer committed:
don't force migration

ikreymer committed:
add revision if profileid is changed

SuaYoo committed:
replace config

ikreymer committed:
fix updating endpoint:
- update profiles: update PROFILE_FILENAME in configmap is profileid provided, clear if ""
- return 'settings_changed' and 'metadata_changed' if either crawl settings or metadata changed
- return if neither is true
- update tests to check settigns/metadata changed return values

ikreymer committed:
ensure profileid == "" and null are equivalent when checking if changed
